cpp/src/aztec/polynomials/polynomial.cpp

```diff
@@ -48,112 +37,63 @@ Polynomial<Fr>::Polynomial(std::string const& filename)
 
     if (other.coefficients_ != nullptr) {
-        memcpy(static_cast<void*>(coefficients_), static_cast<void*>(other.coefficients_), sizeof(Fr) * size_);
+        memcpy(static_cast<void*>(coefficients_), static_cast<void*>(other.coefficients_), sizeof(Fr) * other.size_);
     }
-    zero_memory(max_size_);
+    zero_memory(other.size_, size_);
 }
+
 template <typename Fr>
 Polynomial<Fr>::Polynomial(Polynomial<Fr>&& other) noexcept
-    : mapped_(other.mapped_)
-    , coefficients_(other.coefficients_)
-    , initial_size_(other.initial_size_)
-    , size_(other.size_)
-    , page_size_(other.page_size_)
-    , max_size_(other.max_size_)
-    , allocated_pages_(other.allocated_pages_)
-{
-    ASSERT(page_size_ != 0);
-    other.coefficients_ = 0;
-
-    // Clear other, so we can detect use on free after poly's are put in cache
-    other.clear();
-}
+    : coefficients_(std::exchange(other.coefficients_, nullptr))
+    , size_(std::exchange(other.size_, 0))
+    , mapped_(std::exchange(other.mapped_, false))
+{}
 
 template <typename Fr>
-Polynomial<Fr>::Polynomial(Fr* buf, const size_t initial_size_)
-    : mapped_(false)
-    , coefficients_(buf)
-    , initial_size_(initial_size_)
-    , size_(initial_size_)
-    , page_size_(DEFAULT_SIZE_HINT)
-    , max_size_(initial_size_)
-    , allocated_pages_(0)
+Polynomial<Fr>::Polynomial(Fr* buf, const size_t size_)
+    : coefficients_(buf)
+    , size_(size_)
+    , mapped_(false)
 {}
 
 template <typename Fr>
 Polynomial<Fr>::Polynomial()
-    : mapped_(false)
-    , coefficients_(0)
-    , initial_size_(0)
+    : coefficients_(nullptr)
     , size_(0)
-    , page_size_(DEFAULT_SIZE_HINT)
-    , max_size_(0)
-    , allocated_pages_(0)
+    , mapped_(false)
 {}
 
 template <typename Fr> Polynomial<Fr>& Polynomial<Fr>::operator=(const Polynomial<Fr>& other)
 {
-    ASSERT(page_size_ != 0);
     mapped_ = false;
-    initial_size_ = other.initial_size_;
-
-    // Set page size first so that if we do copy from other we allocate according to
-    // other's page size not ours.
-    page_size_ = other.page_size_;
-
-    if (other.max_size_ > max_size_) {
-        // Bump memory and set max_size, before we set size otherwise we will copy an
-        // inappropriate amount of data.
-        bump_memory(other.max_size_);
-    } else {
-        max_size_ = other.max_size_;
-        allocated_pages_ = other.allocated_pages_;
-    }
-
-    size_ = other.size_;
 
-    ASSERT(max_size_ >= size_);
+    ASSERT(other.size_ <= size_);
 
-    if (other.coefficients_ != 0) {
+    if (other.coefficients_ != nullptr) {
         ASSERT(coefficients_);
-        memcpy(static_cast<void*>(coefficients_), static_cast<void*>(other.coefficients_), sizeof(Fr) * size_);
+        memcpy(static_cast<void*>(coefficients_), static_cast<void*>(other.coefficients_), sizeof(Fr) * other.size_);
     }
 
-    zero_memory(max_size_);
+    zero_memory(other.size_, size_);
 
     return *this;
 }
```

Comment:
Yeah I debated about this one. We don't currently do this anywhere and I don't feel we should need to. (The closest thing is something like `barretenberg::polynomial my_poly; my_poly = polynomial(n)` which will use the move assignment operator. Even that should not be necessary but it happens in several places and I didn't want to root it out. Do you see some reason why we need to have an container of empty polynomials rather than constructing polynomials in place in an empty container?

Reply:
I've added functionality to handle (non-move) assignment to a default constructed polynomial